vsphere_plugin_common/__init__.py

```diff
@@ -1793,72 +1856,111 @@ def create_server(
                 customspec.globalIPSettings = globalip
 
                 clonespec.customization = customspec
-        logger().info(
+        self._logger.info(
             'Cloning {server} from {template}.'.format(
                 server=vm_name, template=template_name))
-        logger().debug('Cloning with clonespec: {spec}'
-                       .format(spec=repr(clonespec)))
+        self._logger.debug('Cloning with clonespec: {spec}'
+                           .format(spec=repr(clonespec)))
         task = template_vm.obj.Clone(folder=destfolder,
                                      name=vm_name,
                                      spec=clonespec)
         try:
-            logger().debug(
+            self._logger.debug(
                 "Task info: {task}".format(task=repr(task)))
             if enable_start_vm:
-                logger().info('VM created in running state')
+                self._logger.info('VM created in running state')
                 self._wait_vm_running(task, adaptermaps, os_type == "other")
             else:
-                logger().info('VM created in stopped state')
+                self._logger.info('VM created in stopped state')
                 self._wait_for_task(task)
         except task.info.error:
             raise NonRecoverableError(
                 "Error during executing VM creation task. VM name: \'{0}\'."
                 .format(vm_name))
 
-        vm = self._get_obj_by_name(
+        # VM object created. Now perform final post-creation tasks
+        vm = task.info.result
+        self._logger.info('VM version: vmx-{old}/vmx-{new}'.format(
+            old=self._get_virtual_hardware_version(vm),
+            new=minimal_vm_version))
+        if self._get_virtual_hardware_version(vm) < minimal_vm_version:
+            if enable_start_vm:
```

Comment:
Is there no way we can support enabling starting the VM while still allowing upgrading of VM version?
e.g. perform the upgrade, blocking, then start the VM as usual afterwards?

Reply:
It require rewrite `_task_guest_state_is_running`, as task after `vm.obj.PowerOn()` does not have guest property.

```python
        # get new state of vm
        vm = self._get_obj_by_id(
            vim.VirtualMachine,
            task.info.result._moId,
            use_cache=False,
        )
        if enable_start_vm:
            self._logger.info('VM created in running state')
            task = vm.obj.PowerOn()
            self._wait_vm_running(task, adaptermaps, os_type == "other")
            vm = self._get_obj_by_id(
                vim.VirtualMachine,
                task.info.result._moId,
                use_cache=False,
            )
```